### PR TITLE
test(support-tools-page): add testing for support tools page

### PR DESCRIPTION
**Pull request by awaisdar001**
## Add Config for testing for the support tools app.

* Adding Unit tests for Users app
* Adding API error tests
* Moving `isEmail`, `isValidUsername` to a common utils file.
* Addressed Eslint errors
* Move messages to messages.js (Only a few messages have been moved which are used in tests, in future, all can be moved error messages can be moved here).

Adds the following `npm` commands. 
- `npm run debug-test` => for running tests in debug mode.
- `npm run snapshot` => updates the snapshots 

Coverage from 0% => ~23%

https://openedx.atlassian.net/browse/PROD-2178

### CODE REVIEW

File: src/components/__snapshots__/SupportLinksPage.test.jsx.snap [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        href="/users"
+        id="search-users"
+        onClick={[Function]}
+      >
+        Search Users
+      </a>
+    </li>
+  </ul>
+</main>
+`;
```
Comment (DawoudSheraz) [inline src/components/__snapshots__/SupportLinksPage.test.jsx.snap]: I am not sure if this file should be committed. Is there a specific reason for this file?

File: src/components/SupportLinksPage.test.jsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    expect(searchUsersText).toEqual('Search Users');
+  });
+  it('have correct heading', () => {
+    const wrapper = shallow(<SupportLinksPage />);
+    const supportToolsText = wrapper.find('h3').text();
+    expect(supportToolsText).toEqual('Support Tools');
+  });
+  it('matches snapshot', () => {
+    const tree = renderer.create(<SupportLinksPage />).toJSON();
+    expect(tree).toMatchSnapshot();
```
Comment (DawoudSheraz) [inline src/components/SupportLinksPage.test.jsx]: What does this achieve?

File: src/components/__snapshots__/SupportLinksPage.test.jsx.snap (hunk 1 above)
Comment (awaisdar001) [reply src/components/__snapshots__/SupportLinksPage.test.jsx.snap]: yes, this is the snapshot file which would be required for comparing current (at the time of running builds) and expected snapshot (this one)

File: src/components/SupportLinksPage.test.jsx (hunk 2 above)
Comment (awaisdar001) [reply src/components/SupportLinksPage.test.jsx]: `toMatchSnapshot();` this matches the current snapshot with expected (saved) snapshot.

File: src/PageLoading.test.jsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import renderer from 'react-test-renderer';
+import PageLoading from './PageLoading';
+
+describe('<PageLoading />', () => {
+  it('does not render without message', () => {
+    const wrapper = shallow(<PageLoading srMessage="" />);
+
+    expect(wrapper.find('.sr-only')).toHaveLength(0);
+  });
+  it('does not render without message', () => {
```
Comment (DawoudSheraz) [inline src/PageLoading.test.jsx]: both have the same description. 2nd should be `renders correctly with message`

File: src/PageLoading.test.jsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    const message = 'Loading...';
+    const wrapper = shallow(<PageLoading srMessage={message} />);
+    const srElement = wrapper.find('.sr-only');
+
+    expect(srElement).toHaveLength(1);
+    expect(srElement.text()).toEqual(message);
+  });
+  it('renders correctly', () => {
```
Comment (DawoudSheraz) [inline src/PageLoading.test.jsx]: --> snapshot matches correctly

File: src/components/users/SupportLinksPage.test.jsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  it('correctly renders', () => {
+    expect(wrapper.find('li')).toHaveLength(1);
+  });
+  it('have search users link', () => {
+    const searchUsersText = wrapper.find('#search-users').text();
+
+    expect(searchUsersText).toEqual('Search Users');
+  });
+  it('have correct heading', () => {
```
Comment (DawoudSheraz) [inline src/components/users/SupportLinksPage.test.jsx]: has

File: src/components/users/SupportLinksPage.test.jsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  let wrapper;
+
+  beforeEach(() => {
+    wrapper = shallow(<SupportLinksPage />);
+  });
+
+  it('correctly renders', () => {
+    expect(wrapper.find('li')).toHaveLength(1);
+  });
+  it('have search users link', () => {
```
Comment (DawoudSheraz) [inline src/components/users/SupportLinksPage.test.jsx]: has

File: src/users/UserPage.test.jsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  });
+
+  describe('Checking PropTypes', () => {
+    it('Should not throw a warning', () => {
+      const expectedProps = { location };
+      const propsError = checkProps(UserPage, expectedProps);
+
+      expect(propsError).toBeUndefined();
+    });
+  });
```
Comment (DawoudSheraz) [inline src/users/UserPage.test.jsx]: Why is there a separate describe for a single test?

File: src/users/UserPage.test.jsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  //
+  it('Renders Correctly', () => {
```
Comment (DawoudSheraz) [inline src/users/UserPage.test.jsx]: snapshot match

File: src/users/UserSearch.test.jsx [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+      const userSearchwrapper = mount(<UserSearch {...searchProps} />);
+
+      userSearchwrapper.find('button').simulate('click');
+
+      expect(searchProps.searchHandler).toHaveBeenCalledWith(
+        searchProps.userIdentifier,
+      );
+    });
+
+    it('using snapshot', () => {
```
Comment (DawoudSheraz) [inline src/users/UserSearch.test.jsx]: matches snapshot

**Review by DawoudSheraz — APPROVED** [review]
For future reference, ideally, the component where the UI changes are more important should be tested via snapshots. For the simple components, the regular unit tests will be enough.

File: src/users/UserPage.test.jsx (hunk 8 above)
Comment (awaisdar001) [reply src/users/UserPage.test.jsx]: just to organize tests. I have added a separate test for proptypes tests.

File: src/dates/formatDate.test.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import formatDate from './formatDate';
+
+describe('Format Date', () => {
+  it('correctly formats date', () => {
+    const formatedDate = formatDate('2015-09-19T11:00:00');
+
+    expect(formatedDate).toEqual('Sep 19, 2015 11:00 AM');
+  });
+  it('fallback to N/A if no args are passed', () => {
```
Comment (DawoudSheraz) [inline src/dates/formatDate.test.js]: nit: returns N/A if data is not provided

File: src/user-messages/messages.js [hunk 11]
```diff
@@ -0,0 +1,4 @@
+export const USER_IDENTIFIER_INVALID_ERROR = 'The searched username or email is invalid. Please correct the username or email and try again.';
+export const USERNAME_IDENTIFIER_NOT_FOUND_ERROR = 'We couldn\'t find a user with the username "{identifier}".';
+export const USER_EMAIL_IDENTIFIER_NOT_FOUND_ERROR = 'We couldn\'t find a user with the email "{identifier}".';
+export const UNKNOWN_API_ERROR = "There was an error loading this user's data. Check the JavaScript console for detailed errors.";
```
Comment (DawoudSheraz) [inline src/user-messages/messages.js:4]: I don't know the current behavior but is there i18n setup for such texts?

File: src/users/UserPage.test.jsx [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  </MemoryRouter>
+);
+
+// Function to wait until the entire component is fully painted.
+const waitForComponentToPaint = async (wrapper) => {
+  await act(async () => {
+    await new Promise((resolve) => setTimeout(resolve));
+    wrapper.update();
+  });
+};
```
Comment (DawoudSheraz) [inline src/users/UserPage.test.jsx:29]: What does component paint stands for?

File: src/users/UserPage.test.jsx [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  describe('Checking PropTypes', () => {
+    it('does not throw a warning', () => {
+      const expectedProps = { location };
+      const propsError = checkProps(UserPage, expectedProps);
+
+      expect(propsError).toBeUndefined();
+    });
+  });
+  describe('shows expected error alert', () => {
+    const mockAuthApiToThrowError = (code) => {
```
Comment (DawoudSheraz) [inline src/users/UserPage.test.jsx]: mockAuthResponseError or something similar to this.

File: src/users/UserPage.test.jsx [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+      );
+
+      const wrapper = mount(<UserPageWrapper location={location} />);
+      await waitForComponentToPaint(wrapper);
+
+      const alert = wrapper.find('.alert');
+      expect(alert).toHaveLength(1);
+      expect(alert.text()).toEqual(expectedAlert);
+    });
+    it('when user email is not found', async () => {
```
Comment (DawoudSheraz) [inline src/users/UserPage.test.jsx]: I haven't seen async usage inside tests before. Can you share an existing example from edX codebase?

File: src/users/UserSearch.test.jsx [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+      expect(propsError).toBeUndefined();
+    });
+    it('does not throw error with undefined userIdentifier', () => {
+      delete props.userIdentifier;
+      const propsError = checkProps(UserSearch, props);
+
+      expect(propsError).toBeUndefined();
+    });
+    it('does throw error with undefined search handler', () => {
```
Comment (DawoudSheraz) [inline src/users/UserSearch.test.jsx]: throws error. not need for does in the beginning.

File: src/utils/index.js [hunk 16]
```diff
@@ -0,0 +1,9 @@
+const EMAIL_REGEX = '^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$';
+const USERNAME_REGEX = '^[\\w.@_+-]+$';
+
+export const isEmail = (value) => !!(value && value.match(EMAIL_REGEX));
```
Comment (DawoudSheraz) [inline src/utils/index.js:5]: No need for !!. This operation is not affecting the value.

File: src/utils/index.js [hunk 17]
```diff
@@ -0,0 +1,9 @@
...
+
+/* eslint-disable arrow-body-style */
+export const isValidUsername = (searchValue) => {
+  return !!(searchValue && searchValue.match(USERNAME_REGEX));
```
Comment (DawoudSheraz) [inline src/utils/index.js:9]: same here

**Review by DawoudSheraz — COMMENTED** [review]
A few comments/questions to address

File: src/user-messages/messages.js (hunk 11 above)
Comment (awaisdar001) [reply src/user-messages/messages.js:4]: I am not aware of that, and these were not being translated currently. I am also not sure if there are plans adding translations for those strings.

File: src/users/UserPage.test.jsx (hunk 12 above)
Comment (awaisdar001) [reply src/users/UserPage.test.jsx:29]: Ah, it just means that component is fully rendered after all promise based function calls.

File: src/users/UserPage.test.jsx (hunk 14 above)
Comment (awaisdar001) [reply src/users/UserPage.test.jsx]: Here is one example. 
https://github.com/edx/frontend-app-learning/blob/927d424d339a8ecde9c542b446770a82bc7dc3cd/src/instructor-toolbar/InstructorToolbar.test.jsx#L47-L54

File: src/utils/index.js (hunk 16 above)
Comment (awaisdar001) [reply src/utils/index.js:5]: @DawoudSheraz  I have copied the methods as is, we can plan on refactoring in the other ticket. I can add todo here as well.

**Comment by awaisdar001** [comment]
@DawoudSheraz  updated the PR.

**Review by saadyousafarbi — APPROVED** [review]
looks good.

File: src/dates/formatDate.test.js [hunk 18]
```diff
@@ -0,0 +1,13 @@
+import formatDate from './formatDate';
+
+describe('Format Date', () => {
+  it('correctly formats date', () => {
```
Comment (asadazam93) [inline src/dates/formatDate.test.js:4]: Do we need to test the utilities as well? I think the priority should be the React components for now

Comment (awaisdar001) [reply src/dates/formatDate.test.js:4]: Yes, true, I did this to bump to coverage a little bit :) and they were easy enough

**Review by asadazam93 — APPROVED** [review]
Looks good to me

**Pull request merged**